cli/internal/runsummary/spaces.go

```diff
@@ -1,15 +1,259 @@
+const tasksEndpoint = "/v0/spaces/%s/runs/%s/tasks"
+
+// spaceRequest contains all the information for a single request to Spaces
+type spaceRequest struct {
+	method  string
+	url     string
+	body    interface{}
+	makeURL func(self *spaceRequest, r *spaceRun) error // Should set url on self
+	onDone  func(self *spaceRequest, response []byte)   // Handler for when request completes
+}
+
+func (req *spaceRequest) error(msg string) error {
+	return fmt.Errorf("[%s] %s: %s", req.method, req.url, msg)
+}
+
+type spacesClient struct {
+	requests   chan *spaceRequest
+	errors     []error
+	api        *client.APIClient
+	ui         cli.Ui
+	run        *spaceRun
+	runCreated chan struct{}
+	wg         sync.WaitGroup
+	spaceID    string
+	enabled    bool
+}
+
+type spaceRun struct {
 	ID  string
 	URL string
 }
 
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui) *spacesClient {
+	return &spacesClient{
+		api:        api,
+		ui:         ui,
+		spaceID:    spaceID,
+		enabled:    spaceID != "",
+		requests:   make(chan *spaceRequest), // TODO: give this a size based on tasks
+		runCreated: make(chan struct{}, 1),
+		run:        &spaceRun{},
+	}
+}
+
+// Start receiving and processing requests in 8 goroutines
+// There is an additional marker (protected by a mutex) that indicates
+// when the first request is done. All other requests are blocked on that one.
+// This first request is the POST /run request. We need to block on it because
+// the response contains the run ID from the server, which we need to construct the
+// URLs of subsequent requests.
+func (c *spacesClient) start() {
+	if !c.enabled {
+		return
+	}
+
+	// Start an immediately invoked go routine that listens for requests coming in from a channel
+	pending := []*spaceRequest{}
+	firstRequestStarted := false
+
+	// Create a labeled statement so we can break out of the for loop more easily
+
+	// Setup a for loop that goes infinitely until we break out of it
+FirstRequest:
+	for {
+		// A select statement that can listen for messages from multiple channels
+		select {
+		// listen for new requests coming in
+		case req := <-c.requests:
+			// Make the first request right away in a goroutine,
+			// queue all other requests. When the first request is done,
+			// we'll get a message on the other channel and break out of this loop
+			if !firstRequestStarted {
+				firstRequestStarted = true
+				go c.dequeueRequest(req)
+			} else {
+				pending = append(pending, req)
+			}
+			// Wait for c.runCreated channel to be closed and:
+		case <-c.runCreated:
+			// 1. flush pending requests
+			for _, req := range pending {
+				go c.dequeueRequest(req)
+			}
+
+			// 2. break out of the forever loop.
+			break FirstRequest
+		}
+	}
+
+	// and then continue listening for more requests as they come in until the channel is closed
+	for req := range c.requests {
+		go c.dequeueRequest(req)
+	}
+}
+
+func (c *spacesClient) makeRequest(req *spaceRequest) {
+	if !c.enabled {
```

Comment:
Does this need to be here? How would you get this far?

If you're concerned, another option is to have a different, empty struct implement the same methods and use that instead of `spacesClient` if spaces is not enabled. Similar to how the [NoOpOutputWatcher](https://github.com/vercel/turbo/blob/3b22f1cf42ec31857c4924bff01ed3ed1fc61502/cli/internal/runcache/output_watcher.go#L18) works.